### ISSUE TITLE
URI components are not unescaped when accessed

### ISSUE THREAD

**Issue by bcardiff** [crystal-lang/crystal#2098]
If the hostname of the URI uses escape % sequence as `http://hi%21` I expect that the uri host property would be  `hi!` and not `hi%21`.

Surprisingly both Ruby and Crystal agrees nowadays in the opposite:

```
irb(main):001:0> require "uri"
irb(main):002:0> URI.parse("http://hi%21").host
=> "hi%21"
```

``` crystal
# crystal
require "uri"
puts URI.parse("http://hi%21").host # => hi%21
```

Should we change this? In that case, when parsing the uri we should avoid unescaping if not necessary in order to avoid some operations.

**Comment by MakeNowJust** [comment]
If this were, for example:

```crystal
require "uri"

u1 = URI.parse "http://example.com%2fexample.com/hello%3fworld%3dfoo"
p u1.host #=> "example.com/example.com"
p u1.path #=> "hello?world=foo"

# construct URL from `u1`
u2 = URI.parse "#{u1.scheme}://#{u1.host}/#{u1.path}"

# `u2` is not equal to `u1` although constructed from `u1`
p u1.host == u2.host #=> false
p u1.path == u2.path #=> false
```

We are afraid of above case.

And more serious problem is character encoding. If given url encoding and expect encoding were mismatched, it might cause serious vulnerability.

All language I know do not unescape URL components automately. Could you tell me the language doing unescape?

**Comment by MakeNowJust** [comment]
Sorry. Swift and Objective-C (rather than Foundation)'s `NSURLComponents` do unescape automately.

```swift
let comp: NSURLComponents? = NSURLComponents(string: "http://example.com%2fexample.com/hello%3fworld%3dfoo")
print(comp?.path) //=> Optional("/hello?world=foo")
```

**Comment by bcardiff** [comment]
Maybe we just need a better way to build safe URIs, like when building html you end up html-escaping the attributes. Because I understand the nice feeling of building a URI with `"#{u1.scheme}://#{u1.host}/#{u1.path}"` but maybe we are just used to a non safe thing here.

**Comment by straight-shoota** [comment]
This has been fixed in #6323

```cr
URI.parse("http://hi%21").host # => "hi!"
```

The problem @MakeNowJust described is not solvable. We can't treat a string as both URI encoded and unencoded at the same time (or rather: depending on the context it's going to be used).
The (somewhat) correct approach would be to enforce URI encoding on the host:

```cr
u1 = URI.parse "http://example.com%2fexample.com/hello%3fworld%3dfoo"
u2 = URI.parse "#{u1.scheme}://#{URI.encode_www_form(u1.host.not_nil!)}#{u1.path}"
p! u1.host == u2.host #=> true
p! u1.path == u2.path #=> true
```
However, this solution can still break easily, and it is generally advised to simply use the `URI` constructor for this:
```cr
u3 = URI.new(scheme: u1.scheme, host: u1.host, path: u1.path)
p! u1.host == u3.host #=> true
p! u1.path == u3.path #=> true
````
